crates/sui-core/src/transaction_orchestrator.rs

```diff
@@ -259,6 +361,75 @@ where
             in_flight.dec();
         });
 
+        // Check if TransactionDriver should be used for submission
+        if self.td_percentage > 0 {
+            let random_value = rand::thread_rng().gen_range(1..=100);
+            if random_value <= self.td_percentage {
+                debug!("Using TransactionDriver for transaction {:?}", tx_digest);
+                if let Some(td) = &self.transaction_driver {
+                    // Add transaction to WAL log for TransactionDriver path
+                    if self
+                        .pending_tx_log
+                        .write_pending_transaction_maybe(&transaction)
+                        .await
+                        .map_err(|e| {
+                            warn!(?tx_digest, "QuorumDriverInternalError: {e:?}");
+                            QuorumDriverError::QuorumDriverInternalError(e)
+                        })?
+                    {
+                        debug!(
+                            ?tx_digest,
+                            "Added transaction to WAL log for TransactionDriver"
+                        );
+                    }
+
+                    let drive_transaction_future = td.drive_transaction(
+                        SubmitTxRequest {
+                            transaction: request.transaction.clone(),
+                        },
+                        SubmitTransactionOptions {
+                            forwarded_client_addr: client_addr,
+                        },
+                    );
+
+                    let Ok(td_response) =
+                        timeout(WAIT_FOR_FINALITY_TIMEOUT, drive_transaction_future).await
+                    else {
+                        debug!(
+                            ?tx_digest,
+                            "[TransactionDriver] Timeout waiting for transaction finality."
+                        );
+                        self.metrics.wait_for_finality_timeout.inc();
+                        return Err(QuorumDriverError::TimeoutBeforeFinality);
+                    };
+                    add_server_timing("[TransactionDriver] wait_for_finality");
+
+                    drop(_txn_finality_timer);
+                    drop(_wait_for_finality_gauge);
+                    self.metrics.wait_for_finality_finished.inc();
+
+                    match td_response {
+                        Err(e) => {
+                            warn!(?tx_digest, "{e:?}");
+                            return Err(QuorumDriverError::TransactionDriverError(e.to_string()));
```

Comment:
Why don't we also want to clean up the transactions from WAL when Err? Do I understand correctly that WAL is mostly meant to be used for crash recovery?